cmd/auth-provider-gcp/plugin/plugin_test.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package plugin
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/cloud-provider-gcp/cmd/auth-provider-gcp/gcpcredential"
+	credentialproviderapi "k8s.io/kubelet/pkg/apis/credentialprovider"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+)
+
+const (
+	serviceAccountsEndpoint = "/computeMetadata/v1/instance/service-accounts/"
+	defaultEndpoint         = "/computeMetadata/v1/instance/service-accounts/default/"
+	scopeEndpoint           = defaultEndpoint + "scopes"
+	emailEndpoint           = defaultEndpoint + "email"
+	tokenEndpoint           = defaultEndpoint + "token"
+	dummyToken              = "ya26.lots-of-indiscernible-garbage"
+	email                   = "1234@project.gserviceaccount.com"
+	expectedUsername        = "_token"
+)
+
+func hasURL(url string, response *credentialproviderapi.CredentialProviderResponse) bool {
+	if _, ok := response.Auth[url]; ok {
+		return true
+	}
+	return false
+}
+
+func usernameMatches(expectedUsername string, auth credentialproviderapi.AuthConfig) bool {
+	return auth.Username == expectedUsername
+}
+
+func passwordMatches(expectedPassword string, auth credentialproviderapi.AuthConfig) bool {
+	return auth.Password == expectedPassword
+}
+
+func TestContainerRegistry(t *testing.T) {
+	// Taken from from pkg/credentialprovider/gcp/metadata_test.go in k/k
+	registryURL := "container.cloud.google.com"
+	token := &gcpcredential.TokenBlob{AccessToken: dummyToken} // Fake value for testing.
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Only serve the URL key and the value endpoint
+		if scopeEndpoint == r.URL.Path {
```

Comment:
nit: for readability, maybe do something more like:

```
prefix := "/computeMetadata/v1/instance/service-accounts/default/"
switch r.URL.Path {
  case prefix + "scopes":
  ...
}
```
?

Having the string consts separated from the handler code seems to be a minus here.

Reply:
I'm not sure if this was the case when this was back in `plugin_test.go`, but some of the constants are used in multiple tests - would it be worth it to duplicate them to move them closer to the HTTP handler?

Reply:
I was wrong about the constants bit actually, the constants used in that test specifically aren't used anywhere else - fixed.